---
title: "Evaluation of different phyllosphere sample types for parallel metabarcoding of Fungi and Oomycetes in Vitis vinifera"
author: "Falk Behrens"
date: "2022-02-07"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

# Methods
Leaf wash sediments (W) and leaf disks (L) are used to characterize the mycobiome of the Vitis vinifera phyllosphere by metabarcoding. Both sample types were processed from each leaf pool. Eighteen fully expanded Vitis vinifera cv. Müller-Thurgau leaves per leaf pool were collected at two sampling time points (02.06.2020 (N1) and 30.06.2020(N2)) at an experimental vineyard located near Neustadt an der Weinstraße, Germany. At each time point leaves of untreated (K1-3), Sulfur (S1-3) treated, or Sulphur and Copper (Cu1-3) treated vines were sampled in three repetitions from randomized Blocks resulting in 18x2 amplicon libraries. Leaf disks (one per leaf; 18 per leaf pool) were cut before washing. Negative controls for DNA extraction (K) and template free library preparation (BPCR) as well as an artificial mock community containing equal amounts of DNA from twelve different in vitro cultivated fungal species (Mock: Fomitiporia mediterranea; Sterum hirsutum; Bjerkandera adusta; Botytis cinerea; Aureobasidium pullulans; Gibellulopsis nigrescens; Phaeoacremonium angustius; Fusarium avenaceum; Fusarium culmorum; Aspergillus ochraceus; Phaeomoniella clamydospora; Pichia kluyveri) were included. A metabarcoding strategy capable of detecting Fungi and Oomycetes by using the ITS1catta/ITS2ngs (Tedersoo et al. 2017; Tedersoo and Anslan, 2019) primer combination for ITS1 library preparation was applied. Library preparation and sequencing were carried out by AllGenetics & Biology SL (www.allgenetics.eu).
...

# Libraries
```{r Libraries}
library(tidyverse)
library(phyloseq)
library(multcomp)
library(ggpubr)
library(vegan)
library(ranacapa)
library(cowplot)
library(emmeans)
library(ggpmisc)
library(lme4)
library(biostat)
library(DESeq2)
```


# Package versions
```{r Packages}
sessionInfo()
```


# Make directory for figures
```{r directory}
dir.create("Figures")
```


# External Functions
### Filtering by negative controls with run_galan()
##### [cpauvert/phyloseq_galan.R](https://gist.github.com/cpauvert)
```{r External functions 1}
# Filtering out contaminants with control samples (Galan et al. 2016)
  # Methods: Galan et al. (2016) doi:10.1128/mSystems.00032-16
  # Implementation: Charlie Pauvert
  # 2018-09-21
  
  
  # Visualize rapidly the filtered OTU (or ASV) table
  viz_matrix<-function(m,axes=T, xlab=NULL, ylab=NULL,...){
    mviz<-t(m) > 0
    image( mviz[,rev(1:ncol(mviz))], xaxt= "n", yaxt= "n", col = c("white","black"),useRaster = T,...)
    if(axes){
      axis( 1, at=seq(0,1,length.out=ncol( m ) ), labels= colnames( m ), las= 2 )
      axis( 2, at=seq(0,1,length.out=nrow( m ) ), labels= rev(rownames( m )), las= 2)
    }
    if(!is.null(xlab)) mtext(xlab,side = 1)
    if(!is.null(ylab)) mtext(ylab,side = 2)
  }
  
  
  # Run the filters
  run_galan<-function(physeq=NULL, ctrl_smp=NULL, positive=NULL, positive_smp=NULL,plot=TRUE){
    if(any(is.null(physeq),is.null(ctrl_smp))){
      stop("physeq or ctrl_cmp is NULL")
    } else {
      
      message("Threshold T_CC : filtering for cross-contamination")
      # Find the maximum number of sequences for an ASV in the
      # negative and positive controls samples
      if(!is.null(positive) & !is.null(positive_smp)){
        # Check if positives samples exists and ASV are given
        positive<-positive[positive %in% taxa_names(physeq)]
        # Fetch only ASV table of the control samples
        ctrl<-prune_samples(c(ctrl_smp,positive_smp),physeq)
        # # And removed the positive control ASVs
        # ctrl<-prune_taxa(!taxa_names(ctrl) %in% positive,ctrl)
      } else {
        # Fetch only negative control samples
        ctrl<-prune_samples(ctrl_smp,physeq)
      }
      m.ctrl<-as(otu_table( ctrl ), "matrix")
      
      
      # T_CC threshold for each ASV in the control samples
      mar<-ifelse(taxa_are_rows(physeq), 1 , 2)
      threshold.ctrl<-apply(m.ctrl,mar,max)
      
      
      if(!is.null(positive)){
        # Positive control ASVs should not be filtered 
        threshold.ctrl[ names(threshold.ctrl) %in% positive]<-0  
      }
      
      # Define a function to relevel ASV to 0 according to a ASV wise threshold.
      adaptative_threshold<-function(physeq, ASVThreshold){
        # Fetch ASV table
        ASV<-otu_table(physeq)
        # set margin depending on ASV orientation: row or column
        taxMargin<-ifelse( taxa_are_rows(ASV), 1, 2)
        # remove the number of sequences indicated in ASVThreshold
        # by samples for each ASV
        ASV.sweeped<-sweep(ASV, taxMargin, ASVThreshold+1, "-")
        # any ASV abundance negative was hence below the threshold
        # and therefore set to NA
        ASV.sweeped[ which(ASV.sweeped < 0) ]<-NA
        # Re balance the altered abundance by adding the previously
        # removed threshold. Note that NA are not affected
        ASV.sweeped<-sweep(ASV.sweeped, taxMargin, ASVThreshold+1,"+")
        # ASV that are NA are set to 0 and considered absent.
        ASV.sweeped[which(is.na(ASV.sweeped))]<-0
        # Return the corrected ASV table
        otu_table(physeq)<-otu_table(ASV.sweeped,
                                     taxa_are_rows = taxa_are_rows(ASV))
        return(physeq)
      }
      
      # Keep the initial ASV table
      initM<-as(otu_table( physeq ), "matrix")
      # Apply the threshold to the ASV table
      physeq<-adaptative_threshold(physeq, threshold.ctrl)
      # Compute the number of cells where the filter occurred
      finalM<-as(otu_table( physeq ), "matrix")
      # Wherever the value is negative, the filter occurred
      finalM<-finalM - initM < 0
      if(plot){
        # Plot the results
        (viz_matrix(finalM,axes=F,main="Threshold T_CC : filtering for cross-contamination"))
      }
      message(paste("Ranges of the T_CC treshold:", min(threshold.ctrl),max(threshold.ctrl)))
      
      if(!is.null(positive) & !is.null(positive_smp)){
        message("\n\nThreshold T_FA : filtering out incorrectly assigned sequences.")
        positive<-positive[positive %in% taxa_names(physeq)]
        # Identification of the maximal number of sequences of "alien" ASV
        # assigned to environmental samples
        alien.max<-max(otu_table(physeq)[positive,! sample_names(physeq) %in% positive_smp])
        # Find which ASV had the maximum values (because the above drops names)
        alien.max.otu<-rownames(which(otu_table(physeq)[positive,! sample_names(physeq) %in% positive_smp] == alien.max,arr.ind = T))
        message(paste("Max alien sequence number:",alien.max))
        message(paste(" identified with the following ASV:",alien.max.otu))
        
        # False-assignment rate R_fa
        alien.ratio<-alien.max / sum(taxa_sums(physeq)[alien.max.otu])
        message(paste("Alien sequence ratio (R_fa):", scales::percent(alien.ratio)))
        
        # Number of sequences potentially misassigned to their samples for each ASV (
        # Threshold T_fa
        threshold.alien<-taxa_sums(physeq) * alien.ratio
        message(paste("Ranges of the T_FA treshold:", min(threshold.alien),round(max(threshold.alien))))
        
        initM<-as(otu_table( physeq ), "matrix")
        
        # Apply the T_fa threshold        
        physeq<-adaptative_threshold(physeq, threshold.alien)
        
        # Compute the number of cells where the filter occurred
        finalM<-as(otu_table( physeq ), "matrix")
        # Wherever the value is negative, the filter occurred
        finalM<-finalM - initM < 0
        if(plot){
          # Plot the results
          (viz_matrix(finalM,axes=F,main="Threshold T_FA : filtering out incorrectly assigned sequences."))
        }
      }
      message("\n\nThe removal of the control samples and the control ASV is NOT done within this function!")
      message("\n\nPlease cite Galan et al. (2016) doi:10.1128/mSystems.00032-16")
      return(physeq)
    }
  }
  #
  # Example of use of the function
  # 
  
  # ASV identifiers of the positive control ASV of the strain.
  # marine.strains<-c(rep('Yamadazyma barbieri',3),rep('Candida oceani',2))
  # names(marine.strains)<-c("e3b29fe75256a310e2712678e84612826b584728",
  #                          "8dd49297074e132374a68ef5219244b0fc512826",
  #                          "3723dfe3dec2cdebeef00493920371525866ad55",# Yama
  #                          "af9d85218cd86808f881938898ad184945244cb0",
  #                          "e37aa309d50c33ef3c5b5521540f762e06cf0b10")
  
  # Applying Galan filters using two thresholds T_CC and T_FA
  # 
  # run_one_filtered<-run_galan(physeq = run_one,
  #                    ctrl_smp = c("B.poole","T.PCR"),
  #                    positive = names(marine.strains),
  #                    positive_smp = c("TplusPCR", "Tplus.ePCR"))
  
  # Example of output by the function
  # 
  # Threshold T_CC : filtering for cross-contamination
  # Ranges of the T_CC treshold: 0 158
  # 
  # 
  # Threshold T_FA : filtering out incorrectly assigned sequences.
  # Max alien sequence number: 310
  #  identified with the following ASV: e3b29fe75256a310e2712678e84612826b584728
  # Alien sequence ratio (R_fa): 0.371%
  # Ranges of the T_FA treshold: 0 8946
  # 
  # 
  # The removal of the control samples and the control ASV is NOT done within this function!
  #   
  #   
  #   Please cite Galan et al. (2016) doi:10.1128/mSystems.00032-16
```

### Rarecurves with ggrare() (modified)
##### [ gauravsk/ranacapa](https://github.com/gauravsk/ranacapa)
```{r External functions 2}
# Rareplots
# Linetype and dot arguments were included in for plotting
# Plot rarefaction curves 
ggrare <- function(physeq, step = 10, label = NULL, color = NULL, linetype = NULL, plot = TRUE, parallel = FALSE, se = TRUE, dot = FALSE) {
  ## Args:
  ## - physeq: phyloseq class object, from which abundance data are extracted
  ## - step: Step size for sample size in rarefaction curves
  ## - label: Default `NULL`. Character string. The name of the variable
  ##          to map to text labels on the plot. Similar to color option
  ##          but for plotting text.
  ## - linetype: Default `NULL`. Character string. The name of the variable
  ##          to map to linetypes on the plot. Similar to color option
  ##          but for plotting different linetypes.
  ## - color: (Optional). Default 'NULL'. Character string. The name of the
  ##          variable to map to colors in the plot. This can be a sample
  ##          variable (among the set returned by
  ##          'sample_variables(physeq)' ) or taxonomic rank (among the set
  ##          returned by 'rank_names(physeq)').
  ##
  ##          Finally, The color scheme is chosen automatically by
  ##          'link{ggplot}', but it can be modified afterward with an
  ##          additional layer using 'scale_color_manual'.
  ## - color: Default `NULL`. Character string. The name of the variable
  ##          to map to text labels on the plot. Similar to color option
  ##          but for plotting text.
  ## - plot:  Logical, should the graphic be plotted.
  ## - parallel: should rarefaction be parallelized (using parallel framework)
  ## - se:    Default TRUE. Logical. Should standard errors be computed.
  ## require vegan
  x <- as(otu_table(physeq), "matrix")
  if (taxa_are_rows(physeq)) { x <- t(x) }
  
  ## This script is adapted from vegan `rarecurve` function
  tot <- rowSums(x)
  S <- rowSums(x > 0)
  nr <- nrow(x)
  
  rarefun <- function(i) {
    cat(paste("rarefying sample", rownames(x)[i]), sep = "\n")
    n <- seq(1, tot[i], by = step)
    if (n[length(n)] != tot[i]) {
      n <- c(n, tot[i])
    }
    y <- rarefy(x[i, ,drop = FALSE], n, se = se)
    if (nrow(y) != 1) {
      rownames(y) <- c(".S", ".se")
      return(data.frame(t(y), Size = n, Sample = rownames(x)[i]))
    } else {
      return(data.frame(.S = y[1, ], Size = n, Sample = rownames(x)[i]))
    }
  }
  if (parallel) {
    out <- mclapply(seq_len(nr), rarefun, mc.preschedule = FALSE)
  } else {
    out <- lapply(seq_len(nr), rarefun)
  }
  df <- do.call(rbind, out)
  
  ## Get sample data
  if (!is.null(sample_data(physeq, FALSE))) {
    sdf <- as(sample_data(physeq), "data.frame")
    sdf$Sample <- rownames(sdf)
    data <- merge(df, sdf, by = "Sample")
    labels <- data.frame(x = tot, y = S, Sample = rownames(x))
    labels <- merge(labels, sdf, by = "Sample")
  }
  
  ## Add, any custom-supplied plot-mapped variables
  if( length(color) > 1 ){
    data$color <- color
    names(data)[names(data)=="color"] <- deparse(substitute(color))
    color <- deparse(substitute(color))
  }
  if( length(label) > 1 ){
    labels$label <- label
    names(labels)[names(labels)=="label"] <- deparse(substitute(label))
    label <- deparse(substitute(label))
  }
  
  p <- ggplot(data = data, aes_string(x = "Size", y = ".S", group = "Sample", color = color))
  p <- p + labs(x = "Sample Size", y = "Species Richness")
  if (!is.null(label)) {
    p <- p + geom_text(data = labels, aes_string(x = "x", y = "y", label = label, color = color),
                       size = 4, hjust = 0)
  }
  p <- p + geom_line(mapping = aes_string(linetype = linetype), size = 0.5)
  if (se) { ## add standard error if available
    p <- p + geom_ribbon(aes_string(ymin = ".S - .se", ymax = ".S + .se", color = NULL, fill = color), alpha = 0.2)
  }
  if (!is.null(dot)) {
    p <- p + geom_point(data = labels, aes_string(x = "x", y = "y", fill = dot), shape = 21, color = "black", size = 2)
  }
  if (plot) {
    plot(p)
  }
  invisible(p)
}
```


# Prepare ASV physeqs
### Load metabarcoding and sample data
```{r Load metabarcoding data}
## summarize physeq data
# Set path for working directory
path <- getwd()

# merged tables
table <- sort(list.files(path, pattern = "merged.table", full.names = TRUE))
table

# sampledata
sampletab <- list.files(path, pattern = "samplelist", full.names = TRUE)
sampletab
```
### Make ASV physeq list
```{r Physeq lists, warning=F, message=F}
# Physeq list
ASV <- list()

# load table from DADA2 analysis
ASV[["merged.table"]] <- read.table(table, sep="\t", header = TRUE)

# make sequence tbl to create fasta
ASV[["blast.table"]] <- data.frame(ID = rownames(ASV[["merged.table"]]), Sequence = ASV[["merged.table"]]$Sequence)

# make sampletab
ASV[["sampledata"]] <- sample_data(data.frame(read.table(sampletab, header= TRUE, sep="\t", dec=","), stringsAsFactors=FALSE))

# make basic.physeq
ASV[["basic.physeq"]] <- phyloseq(otu_table(as(ASV[["merged.table"]][ ,c(1:(length(colnames(ASV[["merged.table"]]))-9))], "matrix"), taxa_are_rows = TRUE), tax_table(as(ASV[["merged.table"]][,c((length(colnames(ASV[["merged.table"]]))-8):(length(colnames(ASV[["merged.table"]]))-1))], "matrix")), ASV[["sampledata"]])

# neg.filt.physeq
ASV[["neg.filt.physeq"]] <- run_galan(physeq = ASV[["basic.physeq"]], ctrl_smp = rownames(ASV[["sampledata"]][ASV[["sampledata"]]$Treatment == "negative"]), plot = FALSE)
ASV[["neg.filt.physeq"]] <- prune_samples(!sample_names(ASV[["neg.filt.physeq"]]) %in% rownames(ASV[["sampledata"]][ASV[["sampledata"]]$Treatment == "negative"]), ASV[["neg.filt.physeq"]])
ASV[["neg.filt.physeq"]] <- prune_taxa(taxa_sums(ASV[["neg.filt.physeq"]]) > 0, ASV[["neg.filt.physeq"]])

# mock.physeq
ASV[["mock.physeq"]] <- subset_samples(ASV[["neg.filt.physeq"]], Treatment == "positive")
ASV[["mock.physeq"]] <- prune_taxa(taxa_sums(ASV[["mock.physeq"]]) > 0, ASV[["mock.physeq"]])

# sample.physeq
ASV[["sample.physeq"]] <- subset_samples(ASV[["neg.filt.physeq"]], Sampletype != "Control" )
ASV[["sample.physeq"]] <- prune_taxa( taxa_sums(ASV[["sample.physeq"]]) > 0, ASV[["sample.physeq"]])
```

### Function to filter ASVs based on abundance in n samples
```{r Filter abundance function}
filter_abundance <- function(physeq, quant = 2, samp = 2/36){
  filt = genefilter_sample(physeq, filterfun_sample(function(x) x >= quant), A=(samp)*nsamples(physeq))
  new = prune_taxa(filt, physeq)
  return(new)
}
```

### Filter physeq by ASV occurrence and taxa
```{r Filter physeq}
# Filtering
physeq <- ASV[["sample.physeq"]]
physeq
physeq <- subset_taxa(physeq, Kingdom != "Unknown")
physeq
physeq.filt <- filter_abundance(physeq, samp = 2/36)
physeq.filt

# Fungi and Stramenopila
physeq.fungi <- subset_taxa(physeq.filt, Kingdom == "Fungi" | Kingdom == "Stramenopila")
physeq.fungi
sample_sums(physeq.fungi)
```


# Analyse ASV occurence
### Bar plot for sampletype dependent occurrence of ASVs
```{r Bar plot, fig.height=10, fig.width=8}
otu.tab <- tibble(merge(data.frame(otu_table(physeq.fungi)) %>%
                          dplyr::select(seq(1,36,2)),
                        data.frame(otu_table(physeq.fungi)) %>%
                          dplyr::select(seq(2,36,2)),
                        by = "row.names"))
colnames(otu.tab)[1] <- "ID"
otu.tab <- left_join(otu.tab, data.frame(tax_table(physeq.fungi)),
                     by = "ID")

factor_groups <- seq(1,18,9)
sample.list <- list()

for (i in 1:length(factor_groups)) {
  x <- factor_groups[i] 
  y <- x+1 
  
  df <- tibble(otu.tab[1],
               leaf_disk = rowSums(otu.tab[,seq(y,y+8)]),
               leaf_wash = rowSums(otu.tab[,seq(y+18,y+26)]),
               otu.tab[c(39,44)])
  df[,6:8] <- "XXX"
  colnames(df)[6:8] <- c("sum", "occ", "Timepoint")
  
  # sum 
  df[6] <- df[,2] + df[,3]
  df <- filter(df, sum != 0)
  #df <- filter(df, Phylum == "Ascomycota" | Phylum == "Basidiomycota" | Phylum == "Oomycota" )
  
  # specific occurence
  for (z in 1:nrow(df)){
    if(df[z,2] >= 1 && df[z,3] >= 1){
      df$occ[z] <- "Both"
    } else if(df[z,2] >= 1 && df[z,3] == 0){
      df$occ[z] <- "Leaf disk"
    } else if(df[z,2] == 0 && df[z,3] >= 1){
      df$occ[z] <- "Leaf wash"
    } else {
      print("Error in assignment of specific occurence")
    }
  }
  
  # Timepoint
  if(x == 1){
    df$Timepoint <- "2020_06_02"
  } else if(x == 10){
    df$Timepoint <- "2020_06_30"
  } else {
    print("Error in assign Timepoint")
  }

  sample.list[[i]] <- df
}

for (x in 1:2) {
  by_occ <- group_by(sample.list[[x]], occ, Phylum, Timepoint)
  sample.list[[x]] <- summarise(by_occ,
                                count = n())
}

occ.tab <- bind_rows(sample.list, .id = "sample")

by_occurrence <- group_by(occ.tab, Timepoint, occ)
occ_sum <- data.frame(summarise(by_occurrence,
                       count = sum(count)))

# New lables for Treatment
Treatment.labs <- c("Untreated\ncontrol", "Sulphur", "Sulphur +\nCopper")
names(Treatment.labs) <- c("Cont", "S", "S_Cu")

# New lables for Timepoint
Timepoint.labs <- c("02.06.2020", "30.06.2020")
names(Timepoint.labs) <- c("2020_06_02", "2020_06_30")

p3 <- ggplot(occ.tab, aes(x = factor(occ, levels = c("Leaf disk", "Both", "Leaf wash")), y = count, fill = Phylum))+
  geom_bar(stat = "identity", color = "black")+
  facet_wrap(~Timepoint, labeller = labeller(Timepoint = Timepoint.labs), ncol = 2)+
  
  labs(x="Sample types",
       y="Observed ASVs")+
  scale_fill_manual(name = "Phylum:",
                    values = c("firebrick3", "forestgreen", "chocolate2", "darkorchid3", "dodgerblue3", "gold3"))+
  
  geom_text(data = occ_sum, aes(x = factor(occ, levels = c("Leaf disk", "Both", "Leaf wash")), y = count+10, fill = NULL, label=count),
             color="black", size = 5)+
    
  guides(fill=guide_legend(direction = "vertical", ncol = 2))+
  
  theme_bw()+
  theme(strip.text.x = element_text(size = 12, color = "black", face = "bold"))+
  theme(axis.title.x = element_text(vjust= -1, size=14))+
  theme(axis.text.x = element_text(hjust = 0.5, vjust= 0.5, size = 12))+
  theme(axis.title.y = element_text(hjust = 0.5, vjust= 2, size = 14))+
  theme(axis.text.y = element_text(hjust = 0.5, vjust= 0.5, size = 12))+
  theme(legend.title = element_text(hjust = 0, vjust= 0.5, size = 14))+
  theme(legend.text = element_text(hjust = 0, vjust= 0.5, size = 12))+
  theme(legend.key.size = unit(0.7, "cm"))+
  theme(legend.spacing.x = unit(0.2, "cm"))+
  theme(legend.position= "bottom")

p3

ggsave("./Figures/bar_plot.pdf", p3, width = 5, height = 7)

```

### Leaf disk ASVs
```{r Leaf disk ASVs}
prune_taxa(taxa_sums(subset_samples(physeq.fungi, Sampletype == "Leaf")) > 0, subset_samples(physeq.fungi, Sampletype == "Leaf"))
```

### Leaf wash ASVs
```{r Leaf wash ASVs}
prune_taxa(taxa_sums(subset_samples(physeq.fungi, Sampletype == "Wash")) > 0, subset_samples(physeq.fungi, Sampletype == "Wash"))
```

### Percent of Vitales sequences of Leaf disk samples
```{r  Vitales sequences of Leaf disk samples}
sub_leaf_vitales <- subset_taxa(subset_samples(physeq.filt, Sampletype == "Leaf"), Order == "Vitales")
sum(data.frame(otu_table(sub_leaf_vitales)))/sum(data.frame(otu_table(subset_samples(physeq.filt, Sampletype == "Leaf"))))*100
```

### Percent of Vitales sequences of Leaf wash samples
```{r Vitales sequences of Leaf wash samples}
sub_wash_vitales <- subset_taxa(subset_samples(physeq.filt, Sampletype == "Wash"), Order == "Vitales")
sum(data.frame(otu_table(sub_wash_vitales)))/sum(data.frame(otu_table(subset_samples(physeq.filt, Sampletype == "Wash"))))*100
```

### Counting Phyla Abundances
```{r Phyla Abundances}
tax_tab <- data.frame(tax_table(physeq.filt))
nrow(tax_tab[tax_tab$Phylum == "Ascomycota",])
nrow(tax_tab[tax_tab$Phylum == "Basidiomycota",])
nrow(tax_tab[tax_tab$Phylum == "Oomycota",])
```


# α-Diversity
### Function to make α-Diversity table
```{r α-Diversity function}
make_alpha_table <- function(physeq = physeq){
  
  # Estimate diversity indices
  otu_df <- data.frame(otu_table(physeq))
  df <- data.frame(Sample = colnames(otu_df), row.names = colnames(otu_df))
  
  # Indices
  for (x in 1:ncol(otu_df)) {
    
    # sample column
    col <- otu_df[otu_df[,x] != 0,x]
    
    # Observed
    df$Observed[x] <- length(col)
    
    # Shannon
    value <- 0
    for (z in 1:length(col)){
      value <- value + (col[z]/sum(col))*log(col[z]/sum(col))*-1
    }
    df$Shannon[x] <- value
    
    # Pielou's evenness
    df$Evenness[x] <- df$Shannon[x]/log(df$Observed[x])
    
    # InvSimpson
    value <- 0
    for (z in 1:length(col)){
      value <- value + (col[z]/sum(col))^2
    }
    df$InvSimpson[x] <- 1/value
  }
  df <- merge(df, data.frame(data.frame(sample_data(physeq))), by  = "row.names")
  df$Row.names <- NULL
  return(df)
}
```

### Make α-Diversity table
```{r Alpha diversity table}
alpha_table <- make_alpha_table(physeq.fungi)
alpha_table$Combi <- as.factor(paste(alpha_table$Sampletype, alpha_table$Timepoint, sep = "_"))
alpha_table <- alpha_table[order(alpha_table$Treatment),]
alpha_table <- alpha_table[order(alpha_table$Sampletype),]
alpha_table <- alpha_table[order(alpha_table$Timepoint),]
head(alpha_table)
```

### Plot rarefaction curves
```{r Plot rarefaction curves}
rASV <- ggrare(physeq.fungi, step = 50, label = NULL, color = "Sampletype", linetype = "Timepoint",
               plot = FALSE, parallel = FALSE, se = FALSE, dot = "Treatment")+

  annotate(geom = "text", x = 0.5, y = 195, label = "A)", 
           vjust = 0, hjust = 0, fontface='bold', size = 10)+
  
  scale_x_continuous(limits = c(0,30000))+
  scale_y_continuous(limits = c(0,200))+
  labs(x= "Sequencing depth (reads)", y = "Observed ASVs")+
  
  scale_color_manual(name = "Sample type:",
                     values = c("limegreen", "dodgerblue2"),
                     labels = c("Leaf disk", "Leaf wash"))+
  
  scale_linetype_discrete(name="Sampling time point:",
                          labels = c("02.06.2020", "30.06.2020"))+
  
  scale_fill_manual(name = "Treatment:", 
                    values = c("forestgreen","darkgoldenrod1","cornflowerblue"),
                    labels = c("Untreated\ncontrol", "Sulphur", "Sulphur +\nCopper"))+
  
  guides(color=guide_legend(override.aes = list(size = 1),
                            direction = "horizontal", nrow = 1),
         linetype=guide_legend(override.aes = list(size = 1),
                               direction = "horizontal", nrow = 1),
         fill=guide_legend(override.aes = list(size = 4),
                           direction = "horizontal", nrow = 1))+
  
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5, vjust = 0, size = 16))+
  theme(axis.text.x = element_text(angle = +0, hjust = 0.5, vjust= 0, size = 12))+
  theme(axis.text.y = element_text(angle = +0, hjust = 0, vjust= 0.5, size = 12))+
  theme(axis.title.x = element_text(angle = +0, hjust = 0.5, vjust= 0, size = 14))+
  theme(axis.title.y = element_text(angle = +90, hjust = 0.5, vjust= 2, size = 14))+
  theme(legend.title = element_text(hjust = 0, vjust= 0.5, size = 14))+
  theme(legend.text = element_text(hjust = 0, vjust= 0.5, size = 12))+
  theme(legend.background = element_blank())+
  theme(legend.key.size = unit(1, "cm"))+
  theme(legend.spacing.y = unit(-0.2, "cm"))+
  theme(legend.position= c(0.55,0.125))
```

### GLMM of ASV richness
```{r GLMM of ASV richness}
glmm.mod <- glmer.nb(Observed ~ Sampletype*(Timepoint+Treatment)+(1|Block),
                     data = alpha_table)

# Check model
alpha_table$resid.values <- resid(glmm.mod, type = "working")
alpha_table$fitted.values <- fitted(glmm.mod)

p1 <- ggplot(alpha_table, aes(x=resid.values, fill = Sampletype))+
  geom_histogram(color = "black", bins = 20)+
  labs(title="Residuals")
p2 <- ggplot(alpha_table, aes(x=fitted.values, y=resid.values, fill = Sampletype))+
  geom_hline(aes(yintercept = 0))+
  geom_point(shape = 21)+
  scale_shape_manual(values = c(21,22))+
  labs(title="Residuals")
p3 <- ggplot(alpha_table, aes(sample=resid.values))+
  stat_qq(shape = 21)+
  stat_qq_line(col = "red")+
  labs(title="Normal Q-Q Plot",
       y= "sample quantiles",
       x= "theoretical quantiles")
p4 <- ggplot(alpha_table, aes(x = Sampletype, y=resid.values, fill = Sampletype))+
  geom_boxplot()+
  labs(title="Boxplots of Residuals")
ggarrange(p1,p2,p3,p4, common.legend = TRUE, legend = "bottom")


# ANOVA
car::Anova(glmm.mod)
aov_tab <- round(car::Anova(glmm.mod), digits = 3)
row.names(aov_tab) <- gsub("Sampletype", "Sample type", row.names(aov_tab))
row.names(aov_tab) <- gsub("Timepoint", "Time point", row.names(aov_tab))
row.names(aov_tab) <- gsub(":", " : ", row.names(aov_tab))
for (x in 1:nrow(aov_tab)) {
  if (as.numeric(aov_tab$`Pr(>Chisq)`[x]) <= 0.001){
    aov_tab$`Pr(>Chisq)`[x] <- paste("<0.001")
  }
}

# Tukey test
CIs=cld(emmeans(glmm.mod, ~Sampletype+Timepoint, type = "response"), sort = FALSE, Letters=letters)	
CIs$.group =gsub(" ", "", CIs$.group, fixed = TRUE)	
CIs$.group <- toupper(CIs$.group)
CIs$CL.diff <- CIs$asymp.UCL-CIs$asymp.LCL
CIs
```

### Plot ASV richness
```{r Plot ASV richness}
obs.plot <- ggplot(alpha_table, aes(x = Timepoint, y = Observed))+
  geom_boxplot(aes(fill = Sampletype), position = position_dodge(width = 0.8), width = 0.5, alpha = 0.8)+
  geom_point(aes(color = Sampletype), shape = 21,
             fill = rep(c(rep("forestgreen",3),
                        rep("darkgoldenrod1",3),
                        rep("cornflowerblue",3)),4),
             position = position_jitterdodge(jitter.width = 0.15, dodge.width = 0.8, seed = 142),
             size = 2)+
  geom_point(aes(x = Timepoint, y = 1000, shape = factor(Treatment, levels = c("Cont", "S", "S_Cu"))),
             size = 2)+

  geom_text(x = c(0.8,1.2,1.8,2.2), y = c(max(alpha_table[alpha_table$Combi == "Leaf_2020_06_02",2]) + 10,
                                          max(alpha_table[alpha_table$Combi == "Wash_2020_06_02",2]) + 10,
                                          max(alpha_table[alpha_table$Combi == "Leaf_2020_06_30",2]) + 10,
                                          max(alpha_table[alpha_table$Combi == "Wash_2020_06_30",2]) + 10),
            aes(label = .group), data = CIs, fontface='bold', size = 6)+
  
  annotate(geom = "text", x = 0.5, y = 195, label = "B)", 
           vjust = 0, hjust = 0, fontface='bold', size = 10)+
  annotate(geom = "text", x = 1.37, y = 200, label = "Analysis of Deviance Table (Wald chi-square tests)", 
           vjust = 0, hjust = 0)+
  annotate(geom = "table", x = 1.37, y = 155, label = list(aov_tab), 
           vjust = 0, hjust = 0, table.rownames = TRUE, table.theme = ttheme_gtminimal(padding = unit(c(1, 0.6), "char")))+
  
  scale_y_continuous(limits = c(0,200))+
  scale_x_discrete(labels = c("02.06.2020", "30.06.2020"))+
  scale_color_manual(name = "Sample type:", 
                    values = c("black", "black"),
                    labels = c("Leaf disk", "Leaf wash"),
                    guide = guide_legend(override.aes = list(shape = 21)))+
  scale_shape_manual(name = "Treatment:", 
                     values = c(21, 21, 21),
                     labels = c("Untreated\ncontrol", "Sulphur", "Sulphur +\nCopper"),
                     guide = guide_legend(override.aes = list(size = 4, fill = c("forestgreen","darkgoldenrod1","cornflowerblue")),
                                          direction = "horizontal", nrow = 1))+
  scale_fill_manual(name = "Sample type:", 
                    values = c("limegreen", "dodgerblue2"),
                    labels = c("Leaf disk", "Leaf wash"),
                    guides(fill=guide_legend(direction = "horizontal", nrow = 1)))+
  
  labs(#title = "",
       x= "Sampling time point", y = "")+

  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5, vjust = 0, size = 16))+
  theme(axis.text.x = element_text(angle = +0, hjust = 0.5, vjust= 0, size = 12))+
  theme(axis.text.y = element_text(angle = +0, hjust = 0, vjust= 0.5, size = 12))+
  theme(axis.title.x = element_text(angle = +0, hjust = 0.5, vjust= 0, size = 14))+
  theme(axis.title.y = element_text(angle = +90, hjust = 0.5, vjust= 2, size = 14))+
  theme(legend.title = element_text(hjust = 0, vjust= 0.5, size = 14))+
  theme(legend.text = element_text(hjust = 0, vjust= 0.5, size = 12))+
  theme(legend.background = element_blank())+
  theme(legend.key.size = unit(1, "cm"))+
  theme(legend.spacing = unit(-0.2, "cm"))+
  theme(legend.position= c(0.5,0.125))+
  theme(legend.direction = "horizontal")
```

### Combine α-Diversity plots
```{r α-Diversity plots, fig.height=8, fig.width=16}
rich.plot <- plot_grid(rASV, obs.plot, ncol=2)
rich.plot

ggsave("./Figures/rich_plot.pdf", rich.plot, width = 14, height = 7)

```


# β-Diversity
### Calculate Jaccard distances and PERMANOVA
```{r Jaccard distances and PERMANOVA}
factor.df <- as(sample_data(physeq.fungi), "data.frame")
factor.df$var1 <- row.names(factor.df)
factor.df$var2 <- row.names(factor.df)
dist.mat <- vegan::vegdist(t(as(otu_table(physeq.fungi),"matrix")),method = "jaccard", binary = TRUE)

set.seed(123)
permanova <- adonis(dist.mat ~ Sampletype * (Timepoint + Treatment), factor.df, strata=factor.df$Block)
permanova

perma_tab <- round(permanova[["aov.tab"]], digits = 3)
perma_tab <- perma_tab[,c(1,4,5,6)]
row.names(perma_tab) <- gsub("Sampletype", "Sample type", row.names(perma_tab))
row.names(perma_tab) <- gsub("Timepoint", "Time point", row.names(perma_tab))
row.names(perma_tab) <- gsub(":", " : ", row.names(perma_tab))
for (x in 1:(nrow(perma_tab)-2)) {
  if (as.numeric(perma_tab$`Pr(>F)`[x]) <= 0.001){
    perma_tab$`Pr(>F)`[x] <- paste("<0.001")
  }
}
perma_tab[is.na(perma_tab) == TRUE]  <- ""
```

### NMDS plot
```{r NMDS}
set.seed(123)
nmds <- metaMDS(dist.mat, k = 2, trymax = 100)
stressplot(nmds)

nmds.tab <- merge(scores(nmds), factor.df[,1:4], by= "row.names")
nmds.tab$Combi <- paste(nmds.tab$Sampletype, nmds.tab$Timepoint, sep = "_")

Leaf_2020_06_02 <- nmds.tab[nmds.tab$Combi == "Leaf_2020_06_02", ][chull(nmds.tab[nmds.tab$Combi == "Leaf_2020_06_02", c("NMDS1", "NMDS2")]), ]
Wash_2020_06_02 <- nmds.tab[nmds.tab$Combi == "Wash_2020_06_02", ][chull(nmds.tab[nmds.tab$Combi == "Wash_2020_06_02", c("NMDS1", "NMDS2")]), ]
Leaf_2020_06_30 <- nmds.tab[nmds.tab$Combi == "Leaf_2020_06_30", ][chull(nmds.tab[nmds.tab$Combi == "Leaf_2020_06_30", c("NMDS1", "NMDS2")]), ]
Wash_2020_06_30 <- nmds.tab[nmds.tab$Combi == "Wash_2020_06_30", ][chull(nmds.tab[nmds.tab$Combi == "Wash_2020_06_30", c("NMDS1", "NMDS2")]), ]

hull.data <- rbind(Leaf_2020_06_02, Wash_2020_06_02, Leaf_2020_06_30, Wash_2020_06_30)
hull.data$fill <- "color"
hull.data$fill[hull.data$Sampletype == "Wash"] <- "dodgerblue2"
hull.data$fill[hull.data$Sampletype == "Leaf"] <- "limegreen"

nmds.plot <- ggplot(mapping = aes(x = NMDS1, y = NMDS2))+
  geom_polygon(data=hull.data, aes(color = Combi, group = Combi),alpha = 0.8, fill = hull.data$fill)+
  geom_point(data = nmds.tab,
             mapping = aes(fill = factor(Treatment, levels = c("Cont", "S", "S_Cu")),
                           shape = factor(Combi, levels = c("Leaf_2020_06_02", "Leaf_2020_06_30", "Wash_2020_06_02", "Wash_2020_06_30"))),
             size = 3)+
  annotate(geom = "text", x = -1, y = 0.95, label = "A)", 
           vjust = 0, hjust = 0, fontface='bold', size = 10)+
  annotate(geom = "text", x = -0.2, y = 1, label = "Permutational Multivariate Analysis of Variance Table", 
           vjust = 0, hjust = 0)+
  annotate(geom = "table", x = -0.25, y = 0.41, label = list(perma_tab), 
           vjust = 0, hjust = 0, table.rownames = TRUE, table.theme = ttheme_gtminimal(padding = unit(c(1, 0.6), "char")))+
  annotate(geom = "text", x = -1, y = 0.85, label = paste("  2D stress = ", round(nmds[["grstress"]], digits = 4)), 
           vjust = 0, hjust = 0, size = 4)+

  scale_x_continuous(limits = c(-1,1))+
  scale_y_continuous(limits = c(-1,1))+
  scale_shape_manual(name = "Sample type and\nsampling time point:",
                     labels = c("Leaf disks - 02.06.2020", "Leaf disks - 30.06.2020", "Leaf wash - 02.06.2020", "Leaf wash - 30.06.2020"),
                     values = c(22,23,24,25))+
  scale_fill_manual(name = "\nTreatment:", 
                    values = c("forestgreen", "darkgoldenrod1", "cornflowerblue"),
                    labels = c("Untreated\ncontrol", "Sulphur", "Sulphur + Copper"))+
  scale_color_manual(name = "Sample type and\nsampling time point:",
                     labels = c("Leaf disks - 02.06.2020", "Leaf disks - 30.06.2020", "Leaf wash - 02.06.2020", "Leaf wash - 30.06.2020"),
                     values = rep("black",4))+
  guides(fill=guide_legend(override.aes=list(shape=21, size = 5),
                           direction = "vertical", order = 2, ncol = 1),
         shape=guide_legend(direction = "vertical", order = 1, ncol = 1),
         color=guide_legend(override.aes=list(fill = c("limegreen", "limegreen", "dodgerblue2", "dodgerblue2"),
                                              shape=c(0,5,2,6)),
                            direction = "vertical", order = 1, ncol = 1))+
   
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5, vjust = 0, size = 16))+
  theme(axis.text.x = element_text(angle = +0, hjust = 0.5, vjust= 0, size = 12))+
  theme(axis.text.y = element_text(angle = +0, hjust = 0, vjust= 0.5, size = 12))+
  theme(axis.title.x = element_text(angle = +0, hjust = 0.5, vjust= 0, size = 14))+
  theme(axis.title.y = element_text(angle = +90, hjust = 0.5, vjust= 2, size = 14))+
  theme(legend.title = element_text(hjust = 0, vjust= 0.5, size = 14))+
  theme(legend.text = element_text(hjust = 0, vjust= 0.5, size = 12))+
  theme(legend.background = element_blank())+
  theme(legend.key.size = unit(0.7, "cm"))+
  theme(legend.spacing.x = unit(0.2, "cm"))+
  theme(legend.position= c(0.5,0.14))+
  theme(legend.box = "horizontal")
```

### Within-group Jaccard distances
```{r Plot distances}
# Prepare table for Within-group Jaccard distances
dist.list <- list()
for (i in 1:(nrow(factor.df)-1)) {
  
  x = nrow(factor.df)-i
  
  if(i == 1){
    z = x
  }else if(i > 1){
    z = z+x
  }
  
  if(i == 1){
    y = nrow(factor.df)-z
  }else if(i > 1){
    y = y+x+1
  }
  df <- data.frame(var1 = rep(row.names(factor.df)[i],((nrow(factor.df)-1)-(i-1))),
                   var2 = row.names(factor.df)[-(1:i)],
                   value = dist.mat[y:z])
  df <- merge(df, factor.df, by = "var1", all.x = TRUE)
  df[,length(df)] <- NULL
  colnames(df)[2] <- "var2"
  df <- merge(df, factor.df, by = "var2", all.x = TRUE)
  
  dist.list[[i]] <- df
  
}

dist.tab <- bind_rows(dist.list)
dist.tab <- dist.tab[dist.tab$Sampletype.x == dist.tab$Sampletype.y,]
dist.tab <- dist.tab[dist.tab$Timepoint.x == dist.tab$Timepoint.y,]

row.names(dist.tab) <- 1:nrow(dist.tab)
dist.tab <- dist.tab[,1:7]
colnames(dist.tab) <- c("var2", "var1", "value", "Sampletype", "Timepoint", "Treatment", "Block")
dist.tab$Combi <- paste(dist.tab$Sampletype, dist.tab$Timepoint, sep = "_")

# Pairwise Wilcoxon Rank Sum Tests
wilcox <- pairwise.wilcox.test(dist.tab$value, dist.tab$Combi, p.adjust.method = "BH", paired = TRUE, var.equal = FALSE)
wilcox.tab <- make_cld(wilcox)
wilcox.tab$cld <- toupper(wilcox.tab$cld)

# Plot within-group distances
dist.plot <- ggplot(dist.tab, aes(x = Timepoint, y = value, fill = Sampletype))+
  geom_boxplot(position = position_dodge(width = 0.8), width = 0.5, alpha = 0.8)+
  geom_point(position = position_jitterdodge(dodge.width = 0.8, jitter.width = 0.1))+
  
  annotate(geom = "text", x = 0.5, y = 0.9825, label = "B)", 
           vjust = 0, hjust = 0, fontface='bold', size = 10)+
  
  geom_text(x = c(0.77,1.17,1.77,2.17),
            y = c(max(dist.tab[dist.tab$Combi == "Leaf_2020_06_02",3]) + 0.03,
                  max(dist.tab[dist.tab$Combi == "Wash_2020_06_02",3]) + 0.03,
                  max(dist.tab[dist.tab$Combi == "Leaf_2020_06_30",3]) + 0.03,
                  max(dist.tab[dist.tab$Combi == "Wash_2020_06_30",3]) + 0.03),
            aes(label = cld, fill = NULL), vjust = 0, hjust = 0, 
            data = wilcox.tab, fontface='bold', size = 6)+

  scale_fill_manual(name = "Sample type:", 
                    values = c("limegreen", "dodgerblue2"),
                    labels = c("Leaf disk", "Leaf wash"),
                    guides(fill=guide_legend(direction = "horizontal", nrow = 1)))+
  scale_x_discrete(labels = c("02.06.2020", "30.06.2020"))+
  scale_y_continuous(limits = c(0.25,1))+
  
  labs(x= "Sampling time point", y = "Within-group Jaccard distances")+
  
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5, vjust = 0, size = 16))+
  theme(axis.text.x = element_text(angle = +0, hjust = 0.5, vjust= 0, size = 12))+
  theme(axis.text.y = element_text(angle = +0, hjust = 0, vjust= 0.5, size = 12))+
  theme(axis.title.x = element_text(angle = +0, hjust = 0.5, vjust= 0, size = 14))+
  theme(axis.title.y = element_text(angle = +90, hjust = 0.5, vjust= 2, size = 14))+
  theme(legend.title = element_text(hjust = 0, vjust= 0.5, size = 14))+
  theme(legend.text = element_text(hjust = 0, vjust= 0.5, size = 12))+
  theme(legend.background = element_blank())+
  theme(legend.key.size = unit(1, "cm"))+
  theme(legend.spacing = unit(-0.5, "cm"))+
  theme(legend.position= c(0.5,0.1))+
  theme(legend.direction = "horizontal")
```

### Combine β-Diversity plots
```{r β-Diversity plots, fig.height=8, fig.width=16}
beta.dist.plot <- plot_grid(nmds.plot, dist.plot, ncol=2)
beta.dist.plot

ggsave("./Figures/beta.pdf", beta.dist.plot, width = 14, height = 7)
```

# Abundance of ASVs
#### Only samples with:
#### qPCR measured DNA >= conc. of last valid standard measurement
#### Metabarcoding counts >= 10
### Normalize metabarcoding counts for each sample type 
```{r normalized counts}
# Make metabarcoding count table
dds = phyloseq_to_deseq2(physeq.fungi, ~ Timepoint*Treatment+Block)
dds <- estimateSizeFactors(dds)
normalized_counts <- data.frame(round(counts(dds, normalized=TRUE)))
normalized_counts <- data.frame(t(normalized_counts)) 
```
### Function prepare qPCR data
```{r qPCR function}
qPCR_analysis <- function(file, exclude){
  # make list
  list <- list()
  
  # load and filter data
  tab <- read.table(file, skip = 8, header =TRUE, sep ='\t')
  colnames(tab)[7] <- "Ct"
  tab$Ct <- as.numeric(as.character(tab$Ct))
  for (x in 1:(length(exclude))) {
    tab$Ct[tab$Well == exclude[x]] <- NA
  }
  
  ## prepare standard
  tab_std <- as_tibble(tab[tab$Task == "STANDARD", c(2,3,7,10)])
  by_sample <- group_by(tab_std, Sample.Name)
  tab_std_fixed <- summarise(by_sample,
                             mean_Ct = mean(Ct, na.rm = TRUE),
                             sd_Ct = sd(Ct, na.rm = TRUE),
                             quant = mean(Quantity, na.rm = TRUE))
  for (x in 1:(nrow(tab_std_fixed)-1)) {
    tab_std_fixed[x+1,4] <- tab_std_fixed[x,4]/10
  }
  tab_std_fixed <- tab_std_fixed[!(is.na(tab_std_fixed$mean_Ct)),] 
  tab_std_fixed$log_quant <- log2(tab_std_fixed$quant)
  
  # make model
  std_mod <- lm(log_quant ~ mean_Ct, data = tab_std_fixed)
  tab_std_fixed$pred_quant <- 2^predict(std_mod)
  tab_std_fixed$log_pred_quant <- predict(std_mod)

  # check Std.
  ## log_quant vs. mean_Ct
  std_plot <- ggplot(tab_std_fixed, aes(x = mean_Ct, y = log_quant))+
    geom_smooth(method = "lm")+
    geom_errorbar(aes(xmin=mean_Ct-sd_Ct, xmax=mean_Ct+sd_Ct), size = 0.5, width = 0.25)+
    geom_point(size = 3)+
    stat_regline_equation(label.x=min(tab_std_fixed$mean_Ct), label.y=min(tab_std_fixed$log_quant)+2, size = 6)+
    stat_cor(aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")), label.x=min(tab_std_fixed$mean_Ct), label.y=min(tab_std_fixed$log_quant), size = 6, r.digits = 3)
  
  # Add to list
  list[["Standard"]] <- tab_std_fixed
  list[["Standard.Plot"]] <- std_plot
  list[["Standard.Model"]] <- summary(std_mod)
  
  ## prepare sample table
  tab_samples <- data.frame(tab[tab$Task == "UNKNOWN", c(2,3,7)])
  tab_samples <- na.omit(tab_samples)
  by_sample <- group_by(tab_samples, Sample.Name)
  tab_samples <- summarise(by_sample,
                           tech_rep = n(),
                           mean_Ct = mean(Ct, na.rm = TRUE),
                           sd_Ct = sd(Ct, na.rm = TRUE))
  tab_samples$quantity <- 2^(predict(std_mod, newdata = data.frame(mean_Ct = tab_samples$mean_Ct)))
  tab_samples$log_quantity <- predict(std_mod, newdata = data.frame(mean_Ct = tab_samples$mean_Ct))
  tab_samples <- tab_samples[order(tab_samples$Sample.Name),]
  
  # Add to list  
  list[["Samples"]] <- tab_samples
  
  return(list)
}
```

### Compare Metabarcoding and qPCR
```{r Abundance, fig.height=12, fig.width=12}
# New lables for Sampletypes
Sampletype.labs <- c("Leaf disks", "Leaf wash")
names(Sampletype.labs) <- c("Leaf", "Wash")

# ITS qPCR reference plate 1
ITS1catta_1 <- qPCR_analysis('qpcr_ITS1_Neustadt_2020_leaf_data.txt', c("G7", "H7", "G8", "H8"))
its_qPCR_tab_1 <- ITS1catta_1[["Samples"]]
its_qPCR_tab_1 <- data.frame(ref_quantity = its_qPCR_tab_1$quantity,
                             log_ref_quantity = its_qPCR_tab_1$log_quantity,
                             row.names = its_qPCR_tab_1$Sample.Name)
its_qPCR_tab_1 <- its_qPCR_tab_1[its_qPCR_tab_1$ref_quantity >= ITS1catta_1[["Standard"]][["pred_quant"]][nrow(data.frame(ITS1catta_1[["Standard"]][["pred_quant"]]))],]

# ITS qPCR reference plate 2
ITS1catta_2 <- qPCR_analysis('qpcr_ITS1_Neustadt_2020_wash_data.txt', c("G7", "H7", "G8", "H8"))
its_qPCR_tab_2 <- ITS1catta_2[["Samples"]]
its_qPCR_tab_2 <- data.frame(ref_quantity = its_qPCR_tab_2$quantity,
                             log_ref_quantity = its_qPCR_tab_2$log_quantity,
                             row.names = its_qPCR_tab_2$Sample.Name)
its_qPCR_tab_2 <- its_qPCR_tab_2[its_qPCR_tab_2$ref_quantity >= ITS1catta_2[["Standard"]][["pred_quant"]][nrow(data.frame(ITS1catta_2[["Standard"]][["pred_quant"]]))],]

# Check correlation of standards of both plates
#summary(lm(ITS1catta_1[[1]]$mean_Ct ~ ITS1catta_2[[1]]$mean_Ct))
#summary(lm(ITS1catta_1[[1]]$log_pred_quant ~ ITS1catta_2[[1]]$log_pred_quant))

# Combine data of plates
its_qPCR_tab <- rbind(its_qPCR_tab_1, its_qPCR_tab_2)
its_qPCR_tab <- merge(its_qPCR_tab, data.frame(sample_data(physeq)), by = "row.names")
rownames(its_qPCR_tab) <- its_qPCR_tab$Row.names
its_qPCR_tab$Row.names <- NULL

## Analysis of individual ASVs
# ASV_2
apul_meta_tab <- data.frame(meta_counts = normalized_counts$ASV_2,
                            meta_counts_log = log2((normalized_counts$ASV_2)+1L),
                            row.names = row.names(normalized_counts))
apul_meta_tab <- apul_meta_tab[apul_meta_tab$meta_counts >= 10,]

# ASV_2 qPCR plate 1
apul_1 <- qPCR_analysis('qpcr_Apul_Neustadt_2020_leaf_data.txt', c("G8", "H8"))
apul_qPCR_tab_1 <- apul_1[["Samples"]]
apul_qPCR_tab_1 <- data.frame(target_quantity = apul_qPCR_tab_1$quantity,
                             log_target_quantity = apul_qPCR_tab_1$log_quantity,
                             row.names = apul_qPCR_tab_1$Sample.Name)
apul_qPCR_tab_1 <- apul_qPCR_tab_1[apul_qPCR_tab_1$target_quantity >= apul_1[["Standard"]][["pred_quant"]][nrow(data.frame(apul_1[["Standard"]][["pred_quant"]]))],]

# ASV_2 qPCR plate 2
apul_2 <- qPCR_analysis('qpcr_Apul_Neustadt_2020_wash_data.txt', c("G8", "H8"))
apul_qPCR_tab_2 <- apul_2[["Samples"]]
apul_qPCR_tab_2 <- data.frame(target_quantity = apul_qPCR_tab_2$quantity,
                             log_target_quantity = apul_qPCR_tab_2$log_quantity,
                             row.names = apul_qPCR_tab_2$Sample.Name)
apul_qPCR_tab_2 <- apul_qPCR_tab_2[apul_qPCR_tab_2$target_quantity >= apul_2[["Standard"]][["pred_quant"]][nrow(data.frame(apul_2[["Standard"]][["pred_quant"]]))],]

# Check correlation of standards of both plates
#summary(lm(apul_1[[1]]$mean_Ct ~ apul_2[[1]]$mean_Ct))
#summary(lm(apul_1[[1]]$log_pred_quant ~ apul_2[[1]]$log_pred_quant))

# Combine data of plates
apul_qPCR_tab <- rbind(apul_qPCR_tab_1, apul_qPCR_tab_2)
apul_tab <- merge(apul_qPCR_tab, its_qPCR_tab, by = "row.names")
row.names(apul_tab) <- apul_tab$Row.names
apul_tab$Row.names <- NULL
apul_tab$delta_target_qpcr <- apul_tab$target_quantity / apul_tab$ref_quantity
apul_tab$log_delta_target_qpcr <- log2(apul_tab$delta_target_qpcr)
apul_tab <- merge(apul_tab, apul_meta_tab, by = "row.names")
row.names(apul_tab) <- apul_tab$Row.names
apul_tab$Row.names <- NULL

apul_plot <- ggplot(data = apul_tab, mapping = aes(x = log_delta_target_qpcr, y = meta_counts_log))+
  geom_smooth(method = "lm", color = "black")+
  geom_point(mapping = aes(shape = factor(Timepoint, level = c("2020_06_02", "2020_06_30")), fill = factor(Treatment, level = c("Cont", "S", "S_Cu"))),
             size = 3)+
  facet_grid(~Sampletype, labeller = labeller(Sampletype = Sampletype.labs))+
  stat_regline_equation(label.x=-8, label.y=13.5 ,size = 4)+
  stat_cor(aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")), method = "spearman", cor.coef.name = "rho", label.x=-8, label.y=13 ,size = 4)+
  
  scale_x_continuous(limits = c(-8,0), breaks = seq(-8,0,2))+
  scale_y_continuous(limits = c(6,14), breaks = seq(6,14,2))+
  
  scale_shape_manual(name = "Timepoint:",
                     labels = c("02.06.2020", "30.06.2020"),
                     values = c(21,25))+
  scale_fill_manual(name = "Treatment:", 
                    values = c("forestgreen", "darkgoldenrod1", "cornflowerblue"),
                    labels = c("Untreated\ncontrol", "Sulphur", "Sulphur +\nCopper"))+
  
  labs(title = expression(italic("A. pullulans")* " - ASV 2"),
       x = "", 
       y = expression("Metabarcoding reads [normalized counts (log"[2]*")]"))+

  guides(nrow= 2, fill=guide_legend(override.aes=list(shape=21)))+

  theme_bw()+
  theme(strip.text.x = element_text(size=16))+
  theme(plot.title = element_text(hjust = 0, size = 16))+
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, vjust= 0, size = 12))+
  theme(axis.text.y = element_text(angle = 0, hjust = 0, vjust= 0.5, size = 12))+
  theme(axis.title.x = element_text(angle = 0, hjust = 0.5, vjust= 0, size = 14))+
  theme(axis.title.y = element_text(angle = +90, hjust = 0, vjust= 2, size = 14))+
  theme(legend.position= c("none"))


# ASV_4
eden_meta_tab <- data.frame(meta_counts = normalized_counts$ASV_4,
                            meta_counts_log = log2((normalized_counts$ASV_4)+1L),
                            row.names = row.names(normalized_counts))
eden_meta_tab <- eden_meta_tab[eden_meta_tab$meta_counts >= 10,]

# ASV_4 qPCR plate 1
eden_1 <- qPCR_analysis('qpcr_Eden_Neustadt_2020_leaf_data.txt', c("B1", "B2", "B3", "A4", "A9", "F9",
                                                                   "G8", "H8"))
eden_qPCR_tab_1 <- eden_1[["Samples"]]
eden_qPCR_tab_1 <- data.frame(target_quantity = eden_qPCR_tab_1$quantity,
                             log_target_quantity = eden_qPCR_tab_1$log_quantity,
                             row.names = eden_qPCR_tab_1$Sample.Name)
eden_qPCR_tab_1 <- eden_qPCR_tab_1[eden_qPCR_tab_1$target_quantity >= eden_1[["Standard"]][["pred_quant"]][nrow(data.frame(eden_1[["Standard"]][["pred_quant"]]))],]

# ASV_4 qPCR plate 2
eden_2 <- qPCR_analysis('qpcr_Eden_Neustadt_2020_wash_data.txt', c("G8", "H8"))
eden_qPCR_tab_2 <- eden_2[["Samples"]]
eden_qPCR_tab_2 <- data.frame(target_quantity = eden_qPCR_tab_2$quantity,
                             log_target_quantity = eden_qPCR_tab_2$log_quantity,
                             row.names = eden_qPCR_tab_2$Sample.Name)
eden_qPCR_tab_2 <- eden_qPCR_tab_2[eden_qPCR_tab_2$target_quantity >= eden_2[["Standard"]][["pred_quant"]][nrow(data.frame(eden_2[["Standard"]][["pred_quant"]]))],]

# Check correlation of standards of both plates
#summary(lm(eden_1[[1]]$mean_Ct ~ eden_2[[1]]$mean_Ct))
#summary(lm(eden_1[[1]]$log_pred_quant ~ eden_2[[1]]$log_pred_quant))

# Combine data of plates
eden_qPCR_tab <- rbind(eden_qPCR_tab_1, eden_qPCR_tab_2)
eden_tab <- merge(eden_qPCR_tab, its_qPCR_tab, by = "row.names")
row.names(eden_tab) <- eden_tab$Row.names
eden_tab$Row.names <- NULL
eden_tab$delta_target_qpcr <- eden_tab$target_quantity / eden_tab$ref_quantity
eden_tab$log_delta_target_qpcr <- log2(eden_tab$delta_target_qpcr)
eden_tab <- merge(eden_tab, eden_meta_tab, by = "row.names")
row.names(eden_tab) <- eden_tab$Row.names
eden_tab$Row.names <- NULL


eden_plot <- ggplot(data = eden_tab, mapping = aes(x = log_delta_target_qpcr, y = meta_counts_log))+
  geom_smooth(method = "lm", color = "black")+
  geom_point(mapping = aes(shape = factor(Timepoint, level = c("2020_06_02", "2020_06_30")), fill = factor(Treatment, level = c("Cont", "S", "S_Cu"))),
             size = 3)+
  facet_grid(~Sampletype, labeller = labeller(Sampletype = Sampletype.labs))+
  stat_regline_equation(label.x=-8, label.y=13.5, size = 4)+
  stat_cor(aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")), method = "spearman", cor.coef.name = "rho", label.x=-8, label.y=13, size = 4)+
  
  scale_shape_manual(values = c(21,25))+
  scale_fill_manual(values  = c("forestgreen","darkgoldenrod1", "cornflowerblue"))+
  scale_x_continuous(limits = c(-8,0), breaks = seq(-8,0,2))+
  scale_y_continuous(limits = c(6,14), breaks = seq(6,14,2))+
  
  labs(title = expression(italic("E. dendrobii")* " - ASV 4"),
       x = expression("qPCR measured DNA [ng / ng fungal DNA (log"[2]*")]"), 
       y = expression("Metabarcoding reads [normalized counts (log"[2]*")]"))+
  
  scale_shape_manual(name = "Sampling time point:",
                     labels = c("02.06.2020", "30.06.2020"),
                     values = c(21,25))+
  scale_fill_manual(name = "Treatment:", 
                    values = c("forestgreen", "darkgoldenrod1", "cornflowerblue"),
                    labels = c("Untreated\ncontrol", "Sulphur", "Sulphur +\n Copper"))+
  
  guides(fill=guide_legend(override.aes=list(shape=21, size = 5)),
         shape=guide_legend(override.aes=list(size = 5)),
         nrow= 2)+

  theme_bw()+
  theme(strip.text.x = element_text(size=16))+
  theme(plot.title = element_text(hjust = 0, size = 16))+
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, vjust= 0, size = 12))+
  theme(axis.text.y = element_text(angle = 0, hjust = 0, vjust= 0.5, size = 12))+
  theme(axis.title.x = element_text(angle = 0, hjust = 0.5, vjust= 0, size = 14))+
  theme(axis.title.y = element_text(angle = +90, hjust = 0, vjust= 2, size = 14))+
  theme(legend.title = element_text(hjust = 0, vjust= 0, size = 14))+
  theme(legend.title = element_text(hjust = 0, vjust= 0.5, size = 14))+
  theme(legend.text = element_text(hjust = 0, vjust= 0.5, size = 12))+
  theme(legend.background = element_blank())+
  theme(legend.position= c(0.75,0.15))+
  theme(legend.key.size = unit(0.7, "cm"))+
  theme(legend.spacing.x = unit(0.2, "cm"))+
  theme(legend.spacing.y = unit(0.1, "cm"))+
  theme(legend.direction = "horizontal", legend.box = "vertical")

# Plot ASV_2, ASV_4
plot <- plot_grid(apul_plot, eden_plot, ncol = 1)
plot
ggsave("./Figures/corr_abundance.pdf", plot, width = 10, height = 10)
```


# Citations
```{r}
citation()
citation("tidyverse")
citation("phyloseq")
citation("multcomp")
citation("ggpubr")
citation("vegan")
citation("ranacapa")
citation("cowplot")
citation("emmeans")
citation("ggpmisc")
citation("lme4")
citation("biostat")
citation("DESeq2")
```
















